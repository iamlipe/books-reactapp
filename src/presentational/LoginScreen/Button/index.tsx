import React, {memo} from 'react';
import styled, {useTheme} from 'styled-components/native';

interface Props {
  title: string;
  loading?: boolean;
  onPress: () => void;
}

const Button = ({title, loading = false, onPress}: Props) => {
  const theme = useTheme();

  return (
    <StyledContainer onPress={onPress}>
      {loading ? (
        <StyledLoading size="small" color={theme.colors.PRIMARY} />
      ) : (
        <StyledText>{title}</StyledText>
      )}
    </StyledContainer>
  );
};

const StyledContainer = styled.TouchableOpacity`
  width: 85px;
  height: 35px;

  align-items: center;
  justify-content: center;

  border-radius: 44px;

  background-color: ${({theme}) => theme.colors.WHITE};
`;

const StyledText = styled.Text`
  color: ${({theme}) => theme.colors.PRIMARY};
`;

const StyledLoading = styled.ActivityIndicator``;

export default memo(Button);
