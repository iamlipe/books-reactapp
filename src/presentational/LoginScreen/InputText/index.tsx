import React, {memo, forwardRef} from 'react';
import styled, {useTheme} from 'styled-components/native';
import {ReturnKeyTypeOptions, TextInputProps} from 'react-native';
import {useController} from 'react-hook-form';

interface Props extends TextInputProps {
  name: string;
  control: any;
  label: string;
  placeholder: string;
  secureTextEntry?: boolean;
  returnKeyType?: ReturnKeyTypeOptions | undefined;
  onSubmitEditing?: () => void;
  submitButton?: () => JSX.Element;
}

const InputText = forwardRef(
  (
    {
      name,
      control,
      label,
      placeholder,
      secureTextEntry = false,
      onSubmitEditing = () => null,
      returnKeyType = undefined,
      submitButton: SubmitButton,
    }: Props,
    ref,
  ) => {
    const theme = useTheme();

    const {
      field: {onChange, value},
    } = useController({name, control});

    return (
      <StyledContainer>
        <StyledBackground />
        <StyledContainerInput>
          <StyledRow>
            <StyledColumn>
              <StyledLabel>{label}</StyledLabel>
              <StyledInput
                ref={ref}
                placeholder={placeholder}
                placeholderTextColor={theme.colors.TEXT_LIGHT_TRANSPARENT}
                onChangeText={onChange}
                onSubmitEditing={onSubmitEditing}
                secureTextEntry={secureTextEntry}
                returnKeyType={returnKeyType}
                value={value}
              />
            </StyledColumn>
            {SubmitButton ? <SubmitButton /> : null}
          </StyledRow>
        </StyledContainerInput>
      </StyledContainer>
    );
  },
);

const StyledContainer = styled.View`
  width: 100%;
  height: 60px;

  align-self: center;

  margin-bottom: 16px;
`;

const StyledBackground = styled.View`
  position: absolute;
  top: 0;
  left: 0;

  width: 100%;
  height: 100%;

  border-radius: 4px;

  background-color: ${({theme}) => theme.colors.BLACK};
  opacity: 0.32;
`;

const StyledContainerInput = styled.View`
  height: 100%;

  padding: 6px 16px;
`;

const StyledInput = styled.TextInput`
  flex: 1;

  height: 24px;

  color: ${({theme}) => theme.colors.TEXT_LIGHT};

  padding: 0;
`;

const StyledLabel = styled.Text`
  color: ${({theme}) => theme.colors.TEXT_DARK_TRANSPARENT};

  line-height: 16px;
`;

const StyledRow = styled.View`
  height: 100%;

  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

const StyledColumn = styled.View`
  min-width: 60%;
`;

export default memo(InputText);
