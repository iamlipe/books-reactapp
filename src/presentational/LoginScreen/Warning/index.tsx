import React, {memo} from 'react';
import styled from 'styled-components/native';

import WarningBox from '@assets/svgs/warning-box.svg';

interface Props {
  error: string;
}

const Warning = ({error}: Props) => {
  return (
    <StyledWarningContainer>
      <StyledWarningBox />
      <StyledText>{error}</StyledText>
    </StyledWarningContainer>
  );
};

const StyledWarningContainer = styled.View`
  width: 70%;
  max-width: 239px;
  height: 48px;

  align-items: center;
  justify-content: center;

  padding: 12px 8px 8px 8px;
`;

const StyledWarningBox = styled(WarningBox)`
  position: absolute;
`;

const StyledText = styled.Text`
  color: ${({theme}) => theme.colors.TEXT_LIGHT};
`;

export default memo(Warning);
