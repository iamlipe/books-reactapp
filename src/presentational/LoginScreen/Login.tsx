import React, {useEffect, useRef, useState} from 'react';
import styled from 'styled-components/native';
import * as Yup from 'yup';
import {StatusBar, Keyboard, TextInput} from 'react-native';
import {yupResolver} from '@hookform/resolvers/yup';
import {LoginRequest} from '@store/slices/userSlice';
import {LOGIN} from '@store/slices/userSlice';

import {useForm} from 'react-hook-form';
import {useTranslation} from 'react-i18next';
import {useReduxDispatch} from '@hooks/useReduxDispatch';
import {useReduxSelector} from '@hooks/useReduxSelector';

import Logo from '@assets/svgs/logo.svg';

import backgroud from '@assets/images/backgroud_login.png';

import InputText from './InputText';
import Button from './Button';
import Warning from './Warning';

export const Login = () => {
  const passwordRef = useRef<TextInput>(null);
  const [showError, setShowError] = useState(false);
  const {isLoading, error} = useReduxSelector(state => state.user);
  const {t} = useTranslation();
  const dispatch = useReduxDispatch();

  const schema = Yup.object().shape({
    email: Yup.string()
      .email(t('error.validEmail'))
      .required(t('error.required')),
    password: Yup.string().required(t('error.required')),
  });

  const {
    control,
    handleSubmit,
    formState: {errors},
  } = useForm<LoginRequest>({
    resolver: yupResolver(schema),
  });

  function handleLogin(data: LoginRequest) {
    dispatch(LOGIN(data));
  }

  function onSubmit(data: LoginRequest) {
    handleLogin(data);

    Keyboard.dismiss();
  }

  useEffect(() => {
    if (error) {
      setShowError(true);

      setTimeout(() => setShowError(false), 5000);
    }
  }, [error]);

  return (
    <StyledBackground source={backgroud}>
      <StyledContainer>
        <StyledLogo />
        <InputText
          name="email"
          control={control}
          label={t('label.email')}
          keyboardType="email-address"
          placeholder="books@ioasys.com.br"
          onSubmitEditing={() => passwordRef.current?.focus()}
        />
        <InputText
          ref={passwordRef}
          name="password"
          control={control}
          label={t('label.password')}
          placeholder="********"
          returnKeyType="send"
          onSubmitEditing={handleSubmit(onSubmit)}
          secureTextEntry
          submitButton={() => (
            <Button
              title={t('button.login')}
              onPress={handleSubmit(onSubmit)}
              loading={isLoading}
            />
          )}
        />

        {(showError || errors.email?.message || errors.password?.message) && (
          <Warning
            error={
              errors.email?.message ||
              errors.password?.message ||
              'Email e/ou senha incorretos.'
            }
          />
        )}
      </StyledContainer>
    </StyledBackground>
  );
};

const StyledBackground = styled.ImageBackground`
  min-height: 100%;

  align-items: center;
  justify-content: center;
`;

const StyledLogo = styled(Logo)`
  margin-bottom: 16px;
`;

const StyledContainer = styled.SafeAreaView`
  width: 90%;

  margin-top: ${StatusBar.currentHeight}px;
`;
