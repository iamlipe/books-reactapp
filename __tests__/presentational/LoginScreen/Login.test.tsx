import React from 'react';
import {act, fireEvent} from '@testing-library/react-native';
import {
  renderWithThemeProvider,
  matchSnapshotWithProvider,
} from '@config/functions/Testing';

import {Login} from '@presentational/LoginScreen/Login';

describe('Login', () => {
  test('should render correctly', () => {
    renderWithThemeProvider(<Login />);
  });

  test('should match snapshot', () => {
    matchSnapshotWithProvider(<Login />);
  });

  test('should submit form', async () => {
    const login = renderWithThemeProvider(<Login />);

    const inputEmail = login.getByTestId('input-text-email');
    const inputPassword = login.getByTestId('input-text-password');
    const submitButton = login.getByTestId('submit-button');

    fireEvent.changeText(inputEmail, 'test@email.com');
    fireEvent.changeText(inputPassword, 'passowrd');

    await act(() => fireEvent.press(submitButton));

    expect(inputEmail.props.value).toEqual('test@email.com');
    expect(inputPassword.props.value).toEqual('passowrd');
  });
});
