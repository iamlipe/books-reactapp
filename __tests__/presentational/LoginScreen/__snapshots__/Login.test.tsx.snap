// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match snapshot 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "minHeight": "100%",
      },
    ]
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../src/assets/images/background_login.png",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "marginTop": "px",
          "width": "90%",
        },
      ]
    }
  >
    <SvgMock
      style={
        Array [
          Object {
            "marginBottom": 16,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignSelf": "center",
            "height": 60,
            "marginBottom": 16,
            "width": "100%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#333333",
              "borderRadius": 4,
              "height": "100%",
              "left": 0,
              "opacity": 0.32,
              "position": "absolute",
              "top": 0,
              "width": "100%",
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "height": "100%",
              "paddingBottom": 6,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 6,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "height": "100%",
                "justifyContent": "space-between",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "minWidth": "60%",
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#FFFFFF52",
                    "lineHeight": 16,
                  },
                ]
              }
            >
              Email
            </Text>
            <TextInput
              onChangeText={[Function]}
              onSubmitEditing={[Function]}
              placeholder="books@ioasys.com.br"
              placeholderTextColor="#FFFFFF52"
              secureTextEntry={false}
              style={
                Array [
                  Object {
                    "color": "#FFFFFF",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "height": 24,
                    "paddingBottom": 0,
                    "paddingLeft": 0,
                    "paddingRight": 0,
                    "paddingTop": 0,
                  },
                ]
              }
              testID="input-text-email"
            />
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignSelf": "center",
            "height": 60,
            "marginBottom": 16,
            "width": "100%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#333333",
              "borderRadius": 4,
              "height": "100%",
              "left": 0,
              "opacity": 0.32,
              "position": "absolute",
              "top": 0,
              "width": "100%",
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "height": "100%",
              "paddingBottom": 6,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 6,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "height": "100%",
                "justifyContent": "space-between",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "minWidth": "60%",
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#FFFFFF52",
                    "lineHeight": 16,
                  },
                ]
              }
            >
              Senha
            </Text>
            <TextInput
              onChangeText={[Function]}
              onSubmitEditing={[Function]}
              placeholder="********"
              placeholderTextColor="#FFFFFF52"
              returnKeyType="send"
              secureTextEntry={true}
              style={
                Array [
                  Object {
                    "color": "#FFFFFF",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "height": 24,
                    "paddingBottom": 0,
                    "paddingLeft": 0,
                    "paddingRight": 0,
                    "paddingTop": 0,
                  },
                ]
              }
              testID="input-text-password"
            />
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#FFFFFF",
                "borderRadius": 44,
                "height": 35,
                "justifyContent": "center",
                "opacity": 1,
                "width": 85,
              }
            }
            testID="submit-button"
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#B22E6F",
                  },
                ]
              }
            >
              Entrar
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
